### PR TITLE
Implemented the ACM Stock Market

### PR DESCRIPTION
**Pull request by shravanhariharan2**
Implemented the ACM Stock Market, a new way for members to earn credits. Invest in HACK today! Implemented based on the design doc specs [here](https://www.notion.so/acmucsd/ACM-Stock-Market-Design-Document-49a5a08de3224416a16814cf4c7df6d6)

### CODE REVIEW

**Comment by michl1001** [comment]
_The start of ACMCoin_

File: types/Enums.ts [hunk 1]
```diff
@@ -81,3 +81,16 @@ export enum OrderPickupEventStatus {
...
+export enum StockTransactionType {
+  BUY = 'BUY',
+  SELL = 'SELL',
+}
+
+// need help with these enums...
+export enum StockBroker {
+  SNU_BROKERGE_SERVICES = 'SNU_BROKERAGE_SERVICES',
+  SHAH_BROKERAGE_SERVIES = 'SHAH_BROKERAGE_SERVICES',
+  ROBINHOOD_MARKETS = 'ROBINHOOD_ARKETS',
```
Comment (StormFireFox1) [inline types/Enums.ts:94]: nit: enum has typos
```suggestion
  ROBINHOOD_MARKETS = 'ROBINHOOD_MARKETS',
```

**Review by StormFireFox1 — CHANGES_REQUESTED** [review]
Great work!

So I looked over it, and it works quite well, but I do think we should keep this PR and write some more tests; figure we want to keep the coverage up.

One big issue is I tried investing in HACK, but it says that I don't have enough clout to invest; I thought that was weird, because we both know Dev Team is hella lit.

**Comment by nishantbalaji** [comment]
ACM TO THE MOON

**Review by nishantbalaji — APPROVED** [review]
lgtm 🚀 🚀 🌕

**Comment by UdayKalra** [comment]
Nit: among us

File: api/controllers/StockMarketController.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import { UserAuthentication } from '../middleware/UserAuthentication';
+import { AuthenticatedUser } from '../decorators/AuthenticatedUser';
+import { UserModel } from '../../models/UserModel';
+import StockMarketService from '../../services/StockMarketService';
+import PermissionsService from '../../services/PermissionsService';
+import { CreateStockRequest, StockTransactionRequest } from '../../types';
+
+@UseBefore(UserAuthentication)
+@JsonController('/stock')
+export class StockMarketController {
```
Comment (sumeet-bansal) [inline api/controllers/StockMarketController.ts:11]:
```suggestion
export class StonksMarketController {
```

**Review by sumeet-bansal — APPROVED** [review]
lgtm, quick nit

File: api/controllers/StockMarketController.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  @Post('/')
+  async createStock(@Body() body: CreateStockRequest, @AuthenticatedUser() user: UserModel) {
+    if (!PermissionsService.canCreateStocks(user)) throw new ForbiddenError();
+    const stock = await this.stockMarketService.createStock(body.stock);
+    return { error: null, stock };
+  }
+
+  @Patch('/:symbol')
+  async updateStockPrice(@Body() body: CreateStockRequest, @AuthenticatedUser() user: UserModel) {
```
Comment (sumeet-bansal) [inline api/controllers/StockMarketController.ts:32]: wait what lmao

Comment (sumeet-bansal) [reply api/controllers/StockMarketController.ts:32]: centrally planned capitalism 🚀

**Comment by shravanhariharan2** [comment]
Stakeholders are interested in a more sophisticated stock market pricing algorithm. ACM Stock Market will return...

**Pull request closed without merging**